apollo-router/src/plugins/rhai.rs

```diff
@@ -0,0 +1,507 @@
+};
+use apollo_router_core::{
+    register_plugin, Error, Object, Plugin, RouterRequest, RouterResponse, Value,
+};
+use futures::executor::block_on;
+use http::HeaderMap;
+use http::{header::HeaderName, HeaderValue};
+use rhai::serde::{from_dynamic, to_dynamic};
+use rhai::{Dynamic, Engine, EvalAltResult, Scope, AST};
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde_json_bytes::ByteString;
+
+use tower::{util::BoxService, BoxError, ServiceExt};
+
+const HEADERS_VAR_NAME: &str = "headers";
+const CONTEXT_VAR_NAME: &str = "context";
+
+#[derive(Default, Clone)]
+struct Rhai {
+    ast: AST,
+}
+
+#[derive(Deserialize, JsonSchema)]
+struct Conf {
+    filename: PathBuf,
+}
+
+trait RhaiObjectSetterGetter {
+    fn set(&mut self, key: String, value: Value);
+    fn get_cloned(&mut self, key: String) -> Value;
+}
+
+impl RhaiObjectSetterGetter for Object {
+    fn set(&mut self, key: String, value: Value) {
+        self.insert(ByteString::from(key), value);
+    }
+    fn get_cloned(&mut self, key: String) -> Value {
+        self.get(&ByteString::from(key))
+            .cloned()
+            .unwrap_or_default()
+    }
+}
+
+#[derive(Clone)]
+struct Headers(HeaderMap);
+impl Headers {
+    fn set_header(&mut self, name: String, value: String) -> Result<(), Box<EvalAltResult>> {
+        self.0.append(
+            HeaderName::from_str(&name)
+                .map_err(|err| format!("invalid header name '{name}': {err}"))?,
+            HeaderValue::from_str(&value)
+                .map_err(|err| format!("invalid header value '{value}': {err}"))?,
+        );
+        Ok(())
+    }
+    fn get_header(&mut self, name: String) -> String {
+        self.0
+            .get(&name)
+            .cloned()
+            .map(|h| Some(h.to_str().ok()?.to_string()))
+            .flatten()
+            .unwrap_or_default()
+    }
+}
+
+impl Plugin for Rhai {
+    type Config = Conf;
+
+    fn new(configuration: Self::Config) -> Result<Self, BoxError> {
+        tracing::info!("RHAI {:#?}!", configuration.filename);
+        let engine = Engine::new();
+        let ast = engine.compile_file(configuration.filename)?;
+        Ok(Self { ast })
+    }
+
+    fn router_service(
+        &mut self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        const FUNCTION_NAME: &str = "router_service";
+        if !self
+            .ast
+            .iter_fn_def()
+            .any(|fn_def| fn_def.name == FUNCTION_NAME)
+        {
+            tracing::debug!("RHAI plugin: no router_service function found");
+            return service;
+        }
+        let this = self.clone();
+        let headers_map = Arc::new(Mutex::new(HeaderMap::new()));
+        let headers_cloned = headers_map.clone();
+        let service = service.map_request(move |request: RouterRequest| {
+            let mut headers = headers_cloned.lock().expect("headers lock poisoned");
```

Comment:
a sync Mutex must not be locked from code used in an async context, this can lock the executor. If there's a need for shared state and locking, this should be a proper layer instead of a `map_request` then `map_response`

Reply:
additionally, making it a layer would let us execute the rhai script in a `spawn_blocking`, which will isolate the main executor